clash-lib/src/Clash/Netlist/Id.hs

```diff
@@ -7,111 +7,406 @@
 -}
 
 {-# LANGUAGE CPP #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RecordWildCards #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
 
 module Clash.Netlist.Id
-  ( IdType (..)
-  , mkBasicId'
-  , stripDollarPrefixes
+  ( -- * Utilities to use IdentifierSet
+    IdentifierSet
+  , emptyIdentifierSet
+  , makeSet
+
+    -- * Unsafe creation and extracting identifiers
+  , Identifier
+  , IdentifierType (..)
+  , unsafeMake
+  , toText
+  , toList
+  , union
+
+    -- * Creating and extending identifiers
+  , make
+  , makeBasic
+  , makeBasicOr
+  , makeAs
+  , addRaw
+  , deepen
+  , deepenN
+  , next
+  , nextN
+  , prefix
+  , suffix
+  , fromCoreId
+
+  -- * Misc. and internals
+  , VHDL.stripDollarPrefixes
+  , toBasicId#
+  , isBasic#
+  , isExtended#
   )
 where
 
-import Clash.Annotations.Primitive (HDL (..))
-import Data.Char (isAsciiLower,isAsciiUpper,isDigit)
-import Data.Text as Text
+import           Clash.Annotations.Primitive (HDL (..))
+import           Clash.Core.Name (nameOcc)
+import           Clash.Core.Var (Id, varName)
+import           Clash.Netlist.Types
+  (IdentifierSet(..), Identifier(..), FreshCache, IdentifierType(..),
+   IdentifierSetMonad(identifierSetM))
+import           Control.Arrow (second)
+import qualified Data.Char as Char
+import           Data.Function (on)
+import qualified Data.List as List
+import qualified Data.Text as Text
+import           Data.Text (Text)
+import qualified Data.Maybe as Maybe
+import           Data.Hashable (Hashable(..))
+import           Text.Read (readMaybe)
+import           TextShow (showt)
+import           GHC.Stack (HasCallStack)
 
-data IdType = Basic | Extended
+import qualified Data.IntMap as IntMap
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.HashSet as HashSet
 
-mkBasicId'
-  :: HDL
-  -> Bool
-  -> Text
+import qualified Data.Text.Prettyprint.Doc as PP
+
+import qualified Clash.Netlist.Id.SystemVerilog as SystemVerilog
+import qualified Clash.Netlist.Id.Verilog as Verilog
+import qualified Clash.Netlist.Id.VHDL as VHDL
+import qualified Clash.Netlist.Id.Common as Common
+
+-- | Identifier set without identifiers
+emptyIdentifierSet
+  :: Bool
+  -- ^ Allow escaped identifiers?
+  -> HDL
+  -- ^ HDL to generate names for
+  -> IdentifierSet
+emptyIdentifierSet esc hdl = makeSet esc hdl []
+
+-- | Union of two identifier sets. Errors if given sets have been made with
+-- different options enabled.
+union :: HasCallStack => IdentifierSet -> IdentifierSet -> IdentifierSet
+union is1@(IdentifierSet esc1 hdl1 _ _) is2@(IdentifierSet esc2 hdl2 _ _)
+  | esc1 /= esc2 = error $ "Internal error: esc1 /= esc2, " <> show (esc1, esc2)
+  | hdl1 /= hdl2 = error $ "Internal error: hdl1 /= hdl2, " <> show (hdl1, hdl2)
+  | otherwise = makeSet esc1 hdl1 (toList is1 ++ toList is2)
```

Comment:
I don't think I'm happy with the conversion through list, especially when these `IdentifierSet`s get large. `HashMap` also has a `foldl'`

Reply:
Me neither, but this function should be removed soon(tm). Implementing parallel compile should get rid of the need for it.

Reply:
Functions should still be removed, but `union` is implemented more efficiently now.